### ISSUE TITLE
Logout doesn't properly logout a user

### ISSUE THREAD

**Issue by Jtango18** [Azure/azure-mobile-apps-ios-client#51]
When calling MSClient.logoutWithCompletion, even on a successful logout if the user attempts to login again the login controller picks up their previous details and automatically logs them back in without the opportunity to enter new credentials.

**Comment by phvannor** [comment]
Certain providers will place cookies into the app that will remain, and need to be manually cleared atm.

WIth the use of EasyAuth now, we can have our logout method call the new EA logout (.auth/logout) which will invalidate the Mobile App Service token/cookie.  However, I'm not sure that every provider has a logout itself that would invalidate the provider's cookie.

So you may still have to delete the cookies anyways.    

(note: Sites dating from before EA GA, may have their logout URI stuck at a .auth/<guid> and not the standard /.auth/logout, like most of mine.  We can't ship this until that has been resolved in a Web Apps roll out)

**Comment by phvannor** [comment]
Also, EA logout is currently only functional for AAD users.  Tiwtter/etc would still be valid since it doesn't do anything at the provider level.

**Comment by Jtango18** [comment]
Hi Phillip,

Thanks for the quick response. I expected the issue was with cookies being stored. My log out functionality contains the following code:

     msClient.logoutWithCompletion(){
                error in
                if let error = error{
                    Logger.LogError("Error signing out", error: error)
                }
    
            let storage = NSHTTPCookieStorage.sharedHTTPCookieStorage().cookies
            for cookie in storage!
            {
                NSHTTPCookieStorage.sharedHTTPCookieStorage().deleteCookie(cookie)
            }
            NSUserDefaults.standardUserDefaults().synchronize()
            self.showLoginScreen()

in order to try to counteract this. I still end up with the same problem though. 

I'll look in to calling .auth/logout and see if it resolves the issue.

**Comment by phvannor** [comment]
The cookies get cleared in our e2e testing via: https://github.com/Azure/azure-mobile-apps-ios-client/blob/master/e2etest/ZumoE2ETestApp/ZumoLoginTests.m#L269-L284 and has always seemed to work here for this.

It looks like what you are doing exactly, unless i'm missing something in the ObjC -> Swift conversion in my head. (Except we have that extra filter around, but can't see why that would matter)  

Calling /.auth/logout won't fix this, as we only invalidate the cookie for AAD & Easy Auth itself.  So any of the others would still be passed & valid.

**Comment by Jtango18** [comment]
Close this one off Phillip. It only fails in the simulator, works exactly as expected on a real device. My bad...rookie mistake :-)

**Comment by phvannor** [comment]
Thats good to hear, I'm keeping it open though, as a tracking issue as eventually we need to a) call .auth/logout (hopefully that can call provider's logout in turn) and/or optionally kill the specific cookies for the user to keep it simpler.

**Comment by sachacoder** [comment]
Hi, I have similar problem. I have implemented Azure ADB2C for my Single page application using adal.js library injection. When I logout using /<tenate>/oauth2/v2.0/logout?post_logout_redirect_uri=<appurl>; user is not logged out completely and azure keeps the session. If I open new tab and log in using different user creds, azure sends the auth token of previously logged in user.

Has anyone experienced this problem and provide solution?

**Comment by pjkarels** [comment]
Solution for iOS:

`

            var logoutUrl = "https://login.windows.net/common/oauth2/logout";

            // create safari controller
            var sfViewController = new SFSafariViewController(new NSUrl(logoutUrl), true);

            // show safari controller
            var vc = GetVisibleViewController();

            if (sfViewController.PopoverPresentationController != null)
            {
                sfViewController.PopoverPresentationController.SourceView = vc.View;
            }

            await vc.PresentViewControllerAsync(sfViewController, true);
`

**Comment by ferk0l** [comment]
Does someone have a solution for android? Have tried everything I can think of.

Cleared the Cookies like that:
`    mClient.logout();
    CookieManager cookieManager = CookieManager.getInstance();
    cookieManager.removeAllCookie();`

And also deleted the entire app cache and google chrome cache.

**Comment by pjkarels** [comment]
Something like this. The Chrome Tabs doesn't automatically close and go back to your app, however.
`

            var logoutUrl = "https://login.windows.net/common/oauth2/logout";
            var uri = Android.Net.Uri.Parse(logoutUrl);

            var builder = new CustomTabsIntent.Builder();
            var customTabsIntent = builder.Build();
            customTabsIntent.LaunchUrl(MainActivity, uri);
`
